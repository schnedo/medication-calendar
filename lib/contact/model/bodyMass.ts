export default class BodyMass {
  static readonly unit = "kg";

  constructor(readonly amount: number) {}

  toString = (): string => `${this.amount} ${BodyMass.unit}`;
}
