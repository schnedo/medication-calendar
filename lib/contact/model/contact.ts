export default class Contact {
  constructor(
    readonly fullName: string,
    readonly address: string,
    readonly phoneNumber: string,
  ) {}
}
