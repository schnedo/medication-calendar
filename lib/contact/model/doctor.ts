import Contact from "./contact";

export default class Doctor extends Contact {}
