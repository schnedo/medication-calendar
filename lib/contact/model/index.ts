export type { default as BodyMass } from "./bodyMass";
export { format as formatBodyMass } from "./bodyMass";
export type { default as Doctor } from "./doctor";
export type { default as TreatmentCenter } from "./treatmentCenter";
export type { default as User } from "./user";
