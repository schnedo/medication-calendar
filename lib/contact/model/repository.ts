import Doctor from "./doctor";
import TreatmentCenter from "./treatmentCenter";
import User from "./user";
import * as localforage from "localforage";

const doctorKey = "doctor";
const treatmentCenterKey = "treatmentCenter";
const userKey = "user";

export async function getDoctor(): Promise<Doctor | null> {
  return localforage.getItem(doctorKey);
}

export async function saveDoctor(doctor: Doctor): Promise<Doctor> {
  return localforage.setItem(doctorKey, doctor);
}

export async function getTreatmentCenter(): Promise<TreatmentCenter | null> {
  return localforage.getItem(treatmentCenterKey);
}

export async function saveTreatmentCenter(
  treatmentCenter: TreatmentCenter,
): Promise<TreatmentCenter> {
  return localforage.setItem(treatmentCenterKey, treatmentCenter);
}

export async function getUser(): Promise<User | null> {
  return localforage.getItem(userKey);
}

export async function saveUser(user: User): Promise<User> {
  return localforage.setItem(userKey, user);
}
