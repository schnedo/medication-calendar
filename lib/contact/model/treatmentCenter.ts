import Contact from "./contact";

// There is no export default type ...
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export default interface TreatmentCenter extends Contact {}
