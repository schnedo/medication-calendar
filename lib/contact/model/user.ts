import Contact from "./contact";
import BodyMass from "./bodyMass";

export default interface User extends Contact {
  readonly birthdate: Date;
  readonly bodyMass: BodyMass;
  readonly diagnosis: string;
  readonly fullName: string;
  readonly address: string;
  readonly phoneNumber: string;
}
