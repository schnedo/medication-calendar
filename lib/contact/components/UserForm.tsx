import { ChangeEventHandler, ReactElement, useEffect, useState } from "react";
import {
  Box,
  Button,
  FormGroup,
  makeStyles,
  TextField,
} from "@material-ui/core";
import { DatePicker } from "@material-ui/pickers";
import { User } from "../model";

const useStyles = makeStyles((theme) => ({
  container: {
    "&>*": {
      marginTop: theme.spacing(1),
    },
  },
  buttons: {
    display: "flex",
    width: "100%",
    justifyContent: "flex-end",
  },
}));

interface UserFormState {
  fullName: string;
  birthdate: Date | null;
  address: string;
  phoneNumber: string;
  diagnosis: string;
  bodyMass: string;
}

const transform = ({
  fullName,
  birthdate,
  address,
  phoneNumber,
  diagnosis,
  bodyMass,
}: UserFormState): User => {
  return {
    fullName,
    birthdate,
    address,
    phoneNumber,
    diagnosis,
    bodyMass: bodyMass ? { amount: parseInt(bodyMass) } : null,
  };
};

const defaultState = (user?: User): UserFormState => ({
  address: user?.address ?? "",
  birthdate: user?.birthdate ?? null,
  bodyMass: user?.bodyMass?.amount.toString() ?? "",
  diagnosis: user?.diagnosis ?? "",
  fullName: user?.fullName ?? "",
  phoneNumber: user?.phoneNumber ?? "",
});

export interface UserFormProps {
  user?: User;
  readOnly?: boolean;
  onSubmit?: (value: User) => void;
  onAbort?: () => void;
}

export default function UserForm({
  user,
  readOnly,
  onSubmit,
  onAbort,
}: UserFormProps): ReactElement {
  const [state, setState] = useState<UserFormState>(() => defaultState(user));
  const setField = (
    field: keyof Omit<UserFormState, "birthdate">,
  ): ChangeEventHandler<HTMLInputElement> => ({ target: { value } }) =>
    setState((old) => ({ ...old, [field]: value }));
  const setDate = (birthDate: Date | null) =>
    birthDate && setState((old) => ({ ...old, birthdate: birthDate }));

  useEffect(() => setState(defaultState(user)), [user]);

  const handleAbort = () => {
    setState(defaultState(user));
    onAbort && onAbort();
  };
  const handleSubmit = onSubmit ? () => onSubmit(transform(state)) : undefined;

  const { container, buttons } = useStyles();
  return (
    <FormGroup className={container}>
      <TextField
        value={state.fullName}
        onChange={setField("fullName")}
        id={"fullNameField"}
        label={"Name"}
        variant={"outlined"}
        inputProps={readOnly ? { readOnly: true, disabled: true } : undefined}
      />
      <DatePicker
        value={state.birthdate}
        onChange={setDate}
        id={"birthDatePicker"}
        label={"Geburtstag"}
        inputVariant={"outlined"}
        readOnly={readOnly}
        inputProps={readOnly ? { readOnly: true, disabled: true } : undefined}
      />
      <TextField
        value={state.address}
        onChange={setField("address")}
        id={"addressField"}
        label={"Adresse"}
        variant={"outlined"}
        inputProps={readOnly ? { readOnly: true, disabled: true } : undefined}
      />
      <TextField
        value={state.phoneNumber}
        onChange={setField("phoneNumber")}
        id={"phoneNumberField"}
        label={"Telefonnummer"}
        variant={"outlined"}
        type={"tel"}
        inputProps={readOnly ? { readOnly: true, disabled: true } : undefined}
      />
      <TextField
        value={state.diagnosis}
        onChange={setField("diagnosis")}
        id={"diagnosisField"}
        label={"Diagnose"}
        variant={"outlined"}
        inputProps={readOnly ? { readOnly: true, disabled: true } : undefined}
      />
      <TextField
        value={state.bodyMass}
        onChange={setField("bodyMass")}
        id={"bodyMassField"}
        label={"KÃ¶rpergewicht"}
        variant={"outlined"}
        inputProps={readOnly ? { readOnly: true, disabled: true } : undefined}
      />
      {readOnly ? (
        <></>
      ) : (
        <Box className={buttons}>
          <Button color={"primary"} onClick={handleAbort}>
            Abbrechen
          </Button>
          <Button color={"primary"} onClick={handleSubmit}>
            Speichern
          </Button>
        </Box>
      )}
    </FormGroup>
  );
}
