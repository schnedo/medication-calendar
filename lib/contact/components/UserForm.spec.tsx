/* eslint-disable sonarjs/no-duplicate-string */
import { render } from "@testing-library/react";
import UserForm from "./UserForm";
import { User } from "../model";
import { format } from "date-fns";
import { MuiPickersUtilsProvider } from "@material-ui/pickers";
import DateFnsUtils from "@date-io/date-fns";
import userEvent from "@testing-library/user-event";

describe("UserForm", () => {
  beforeAll(() => {
    jest.setSystemTime(new Date(2011, 1, 12));
  });

  it("should render correctly", async () => {
    expect.hasAssertions();

    const { container } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <UserForm />
      </MuiPickersUtilsProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should get values from user prop if provided", async () => {
    expect.hasAssertions();

    const user1: User = {
      address: "address",
      birthdate: new Date(2021, 1, 10),
      bodyMass: { amount: 56 },
      diagnosis: "diagnosis",
      fullName: "fullName",
      phoneNumber: "phoneNumber",
    };
    const user2: User = {
      address: "address2",
      birthdate: new Date(2021, 2, 20),
      bodyMass: { amount: 57 },
      diagnosis: "diagnosis2",
      fullName: "fullName2",
      phoneNumber: "phoneNumber2",
    };

    const { getByLabelText, rerender } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <UserForm user={user1} />
      </MuiPickersUtilsProvider>,
    );

    const expectDisplayValueEquals = (user: User) => {
      expect(getByLabelText("Name")).toHaveDisplayValue(user.fullName);
      expect(getByLabelText("Geburtstag")).toHaveDisplayValue(
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        format(user.birthdate!, "MMMM do"),
      );
      expect(getByLabelText("Adresse")).toHaveDisplayValue(user.address);
      expect(getByLabelText("Telefonnummer")).toHaveDisplayValue(
        user.phoneNumber,
      );
      expect(getByLabelText("Diagnose")).toHaveDisplayValue(user.diagnosis);
      expect(getByLabelText("Körpergewicht")).toHaveDisplayValue(
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        user.bodyMass!.amount.toString(),
      );
    };

    expectDisplayValueEquals(user1);

    rerender(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <UserForm user={user2} />
      </MuiPickersUtilsProvider>,
    );

    expectDisplayValueEquals(user2);
  });

  it("should be read only when readOnly prop is set", async () => {
    expect.hasAssertions();

    const { getByLabelText, rerender } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <UserForm readOnly />
      </MuiPickersUtilsProvider>,
    );

    let nameField = getByLabelText("Name");
    expect(nameField).toHaveAttribute("readonly");
    expect(nameField).toHaveAttribute("disabled");
    let birthdateField = getByLabelText("Geburtstag");
    expect(birthdateField).toHaveAttribute("readonly");
    expect(birthdateField).toHaveAttribute("disabled");
    let addressField = getByLabelText("Adresse");
    expect(addressField).toHaveAttribute("readonly");
    expect(addressField).toHaveAttribute("disabled");
    let phoneNumberField = getByLabelText("Telefonnummer");
    expect(phoneNumberField).toHaveAttribute("readonly");
    expect(phoneNumberField).toHaveAttribute("disabled");
    let diagnosisField = getByLabelText("Diagnose");
    expect(diagnosisField).toHaveAttribute("readonly");
    expect(diagnosisField).toHaveAttribute("disabled");
    let bodyMassField = getByLabelText("Körpergewicht");
    expect(bodyMassField).toHaveAttribute("readonly");
    expect(bodyMassField).toHaveAttribute("disabled");

    rerender(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <UserForm />
      </MuiPickersUtilsProvider>,
    );

    nameField = getByLabelText("Name");
    expect(nameField).not.toHaveAttribute("readonly");
    expect(nameField).not.toHaveAttribute("disabled");
    birthdateField = getByLabelText("Geburtstag");
    expect(birthdateField).not.toHaveAttribute("disabled");
    addressField = getByLabelText("Adresse");
    expect(addressField).not.toHaveAttribute("readonly");
    expect(addressField).not.toHaveAttribute("disabled");
    phoneNumberField = getByLabelText("Telefonnummer");
    expect(phoneNumberField).not.toHaveAttribute("readonly");
    expect(phoneNumberField).not.toHaveAttribute("disabled");
    diagnosisField = getByLabelText("Diagnose");
    expect(diagnosisField).not.toHaveAttribute("readonly");
    expect(diagnosisField).not.toHaveAttribute("disabled");
    bodyMassField = getByLabelText("Körpergewicht");
    expect(bodyMassField).not.toHaveAttribute("readonly");
    expect(bodyMassField).not.toHaveAttribute("disabled");
  });

  it("should call onSubmit when submitting", async () => {
    expect.hasAssertions();

    const onSubmit = jest.fn();

    const { getByLabelText, getByRole, getByText } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <UserForm onSubmit={onSubmit} />
      </MuiPickersUtilsProvider>,
    );

    await userEvent.type(getByLabelText("Name"), "name");
    userEvent.click(getByLabelText("Geburtstag"));
    userEvent.click(getByRole("button", { name: "10" }));
    userEvent.click(getByRole("button", { name: "OK" }));
    await userEvent.type(getByLabelText("Adresse"), "address");
    await userEvent.type(getByLabelText("Telefonnummer"), "12345");
    await userEvent.type(getByLabelText("Diagnose"), "diagnosis");
    await userEvent.type(getByLabelText("Körpergewicht"), "49");
    userEvent.click(getByText("Speichern"));

    const user: User = {
      address: "address",
      birthdate: new Date(2011, 1, 10),
      bodyMass: { amount: 49 },
      diagnosis: "diagnosis",
      fullName: "name",
      phoneNumber: "12345",
    };
    expect(onSubmit).toHaveBeenCalledTimes(1);
    expect(onSubmit).toHaveBeenLastCalledWith(user);
  });

  it("should call onAbort and reset state when aborting", async () => {
    expect.hasAssertions();

    const onAbort = jest.fn();

    const { getByLabelText, getByRole, getByText } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <UserForm onAbort={onAbort} />
      </MuiPickersUtilsProvider>,
    );

    await userEvent.type(getByLabelText("Name"), "name");
    userEvent.click(getByLabelText("Geburtstag"));
    userEvent.click(getByRole("button", { name: "10" }));
    userEvent.click(getByRole("button", { name: "OK" }));
    await userEvent.type(getByLabelText("Adresse"), "address");
    await userEvent.type(getByLabelText("Telefonnummer"), "12345");
    await userEvent.type(getByLabelText("Diagnose"), "diagnosis");
    await userEvent.type(getByLabelText("Körpergewicht"), "49");
    userEvent.click(getByText("Abbrechen"));

    expect(onAbort).toHaveBeenCalledTimes(1);
    expect(getByLabelText("Name")).toHaveDisplayValue("");
    expect(getByLabelText("Geburtstag")).toHaveDisplayValue("");
    expect(getByLabelText("Adresse")).toHaveDisplayValue("");
    expect(getByLabelText("Telefonnummer")).toHaveDisplayValue("");
    expect(getByLabelText("Diagnose")).toHaveDisplayValue("");
    expect(getByLabelText("Körpergewicht")).toHaveDisplayValue("");
  });
});
