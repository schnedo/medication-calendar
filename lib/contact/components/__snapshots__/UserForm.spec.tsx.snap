// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm should render correctly 1`] = `
<div>
  <div
    class="MuiFormGroup-root makeStyles-container-1"
  >
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        for="fullNameField"
        id="fullNameField-label"
      >
        Name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="fullNameField"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5"
          >
            <span>
              Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root"
      format="MMMM do"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        for="birthDatePicker"
        id="birthDatePicker-label"
      >
        Geburtstag
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="birthDatePicker"
          readonly=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5"
          >
            <span>
              Geburtstag
            </span>
          </legend>
        </fieldset>
      </div>
      
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        for="addressField"
        id="addressField-label"
      >
        Adresse
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="addressField"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5"
          >
            <span>
              Adresse
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        for="phoneNumberField"
        id="phoneNumberField-label"
      >
        Telefonnummer
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="phoneNumberField"
          type="tel"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5"
          >
            <span>
              Telefonnummer
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        for="diagnosisField"
        id="diagnosisField-label"
      >
        Diagnose
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="diagnosisField"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5"
          >
            <span>
              Diagnose
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        for="bodyMassField"
        id="bodyMassField-label"
      >
        Körpergewicht
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="bodyMassField"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5"
          >
            <span>
              Körpergewicht
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-7 makeStyles-buttons-2"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Abbrechen
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Speichern
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
